src/Architectures.jl

```diff
@@ -27,21 +28,49 @@ struct CPU <: AbstractArchitecture end
 """
     GPU <: AbstractArchitecture
 
-Run Oceananigans on a single NVIDIA CUDA GPU.
+Run Oceananigans on a single GPU `device`.
 """
-struct GPU <: AbstractArchitecture end
+struct GPU{D} <: AbstractArchitecture 
+    device :: D
+end
+
+function GPU()
+    if CUDA.has_cuda()
+        return GPU(CUDA.CUDABackend(; always_inline = true))
+    elseif has_metal_device()
+        return GPU(Metal.MetalBackend())
+    else
+        error("No compatible GPU detected")
+    end
+end
+
+"""
+    has_metal_device()
+
+Returns true if a metal device is present (not a current function available from Metal.jl)
+"""
+function has_metal_device()
+    try 
+        Metal.current_device()
+        return true
+    catch
+        return false
+    end
+end
 
 #####
 ##### These methods are extended in DistributedComputations.jl
 #####
 
 device(::CPU) = KernelAbstractions.CPU()
-device(::GPU) = CUDA.CUDABackend(; always_inline=true)
+device(::GPU{<:CUDABackend}) = CUDA.CUDABackend(; always_inline=true)
+device(::GPU{<:MetalBackend}) = Metal.MetalBackend()
```

Comment:
Not sure why we rebuild --- is there reason? try this:

```suggestion
device(gpu::GPU) = gpu.device
```

also useful

```julia
const CUDAGPU = GPU{<:CUDABackend}
const MetalGPU = GPU{<:MetalBackend}
```